### PR TITLE
Refactor FlRenderer to platform-independent implementation

### PR DESCRIPTION
**Pull request by huanghongxun**
This pull request is part of https://github.com/flutter/engine/pull/23985.

This pull request refactors `FlRenderer` to a platform-independent implementation.
1. Switch to embedder compositor.
2. Introduced `FlGLArea` to represent a Flutter content layer from compositor.

Code changes:
1. Removes `egl_utils.h/cc` since I'm using `libepoxy` instead, which is also used by GTK. `mock_egl.cc` is refactored into `mock_epoxy.cc` for mocking `libepoxy`.
2. Drop out `wayland-client` and `x11` requirement. Now we use `gtk_cairo` to draw OpenGL textures.
3. `FlBackingStoreProvider` for framebuffer and texture creation. Just like `FlutterFrameBufferProvider` in macOS implementation.
4. `FlEngine` now enables compositor capability of embedder.
5. `FlGLArea` for framebuffer and texture drawing. every `FlGLArea` has a `GdkWindow`, because we may embed some widgets provided by plugins, which have own `GdkWindow`s.
6. `FlRenderer` is now responsible for compositing. Creates backing store by `FlBackingStoreProvider`, and present layers by putting `FlGLArea`s with given textures into `FlView`.
7. `FlRendererGL` for OpenGL renderer, creates `GdkGLContext` as OpenGL context.
7. `FlView` now is `GtkContainer` to support multiple layers of `FlGLArea`, maintaining a cached `gl_area_list`, and `children_list`. Every frame `FlRenderer` will reset children of `FlView`, but if every time we unparent and parent these child widgets, GtkInspector will crash. So calculates diff of `children_list` is necessary.
8. Pointer events now is captured by a `GtkEventBox` on top of `FlView`, instead of directly capturing events in `FlView`, because `FlGLArea`s have their own `GdkWindow`s, which can capture pointer events, we must cover a `GdkWindow` for event capturing on the top.

This pr is part of platform view implementation on Linux, related issue: flutter/flutter#41724.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

**Comment by huanghongxun** [comment]
@robert-ancell This is the first part of pr to implement platform views on Linux. I have tried to minimize modifications.. But implementing compositor requires all of these changes.

File: shell/platform/linux/fl_backing_store_provider.h
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.h:25]: Missing the start of this sentence?

File: shell/platform/linux/fl_backing_store_provider.cc
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.cc:57]: Change `provider` to `self` (in multiple functions).

File: shell/platform/linux/fl_backing_store_provider.h
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.h:54]: Missing documentation on return value.

File: shell/platform/linux/fl_backing_store_provider.h
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.h:61]: Missing documentation on return value.

File: shell/platform/linux/fl_backing_store_provider.cc
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.cc:81]: This function doesn't need to be public - move the contents into `fl_backing_store_provider_dispose`.

File: shell/platform/linux/fl_backing_store_provider.h
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.h:81]: Function doesn't need to be public.

File: shell/platform/linux/fl_engine.cc
Comment (robert-ancell) [inline shell/platform/linux/fl_engine.cc:196]: This function and the two above need comments above them.

File: shell/platform/linux/fl_gl_area.cc
Comment (robert-ancell) [inline shell/platform/linux/fl_gl_area.cc:28]: Need comments above these functions, e.g. `// Implements GtkWidget::realize`.

File: shell/platform/linux/fl_gl_area.cc
Comment (robert-ancell) [inline shell/platform/linux/fl_gl_area.cc:55]: `area` -> `self`

File: shell/platform/linux/fl_gl_area.cc
Comment (robert-ancell) [inline shell/platform/linux/fl_gl_area.cc:112]: `area` -> `self`

File: shell/platform/linux/fl_renderer_gl.cc
Comment (robert-ancell) [inline shell/platform/linux/fl_renderer_gl.cc:40]: Warning message needs more context, e.g. `"Failed to make make renderer current when creating backing store: %s".

**Comment by huanghongxun** [comment]
> Linux Web Engine
> recipe infra failure: Recipe timed out
All tests are passed now.

File: shell/platform/linux/fl_renderer_gl.cc
Comment (wmww) [inline shell/platform/linux/fl_renderer_gl.cc:25]: Looks sus to use error twice without checking it. probably best to check `error` after each call.

In practice I don't think this is unsafe or a memory leak, but if the 1st call results in an error the 2nd call will fail a non-fatal assertion inside GTK (which prints an error message to console). Might leak an error if assertions are turned off.

**Comment by wmww** [comment]
Tested and Wayland works on Sway and Mir. One oddity is that the GL view seems to be in tiled mode, so when you're resizing the app to be bigger (which is still al little janky) instead of being black, white or whatever was there before, the top and/or right edge is a copy of the other edge side of the window.

File: shell/platform/linux/fl_renderer.h
Comment (huanghongxun) [reply shell/platform/linux/fl_renderer.h:85]: This function implements https://github.com/flutter/engine/blob/0bf71f61401bbe48bc01636ab1a98f447ca57633/shell/platform/embedder/embedder.h#L1050

File: shell/platform/linux/fl_backing_store_provider.cc
Comment (huanghongxun) [reply shell/platform/linux/fl_backing_store_provider.cc:42]: Binding framebuffer 0 does not mean 'unbinding' framebuffer, but binding to the system default framebuffer. Flutter/Skia will always bind a specific framebuffer (provided by `FlBackingStoreProvider`) before rendering. So we have no need to bind to framebuffer 0 here. Restoring framebuffer binding requires a call to `glGetIntegerv(GL_FRAMEBUFFER_BINDING, &old_fbo);`.

macOS implementation has the same behavior, and because framebuffer 0 is not the default framebuffer.
https://github.com/flutter/engine/blob/0bf71f61401bbe48bc01636ab1a98f447ca57633/shell/platform/darwin/macos/framework/Source/FlutterFrameBufferProvider.mm#L31-L37

File: shell/platform/linux/fl_renderer.h
Comment (wmww) [reply shell/platform/linux/fl_renderer.h:85]: Cool

File: shell/platform/linux/fl_backing_store_provider.cc
Comment (huanghongxun) [reply shell/platform/linux/fl_backing_store_provider.cc:14]: Now `geometry` is used in `FlGLArea` for resolving drawing texture repeatedly.

File: shell/platform/linux/fl_renderer_gl.cc
Comment (huanghongxun) [reply shell/platform/linux/fl_renderer_gl.cc:29]: Added error check right after line 24.

**Comment by huanghongxun** [comment]
> Tested and Wayland works on Sway and Mir. One oddity is that the GL view seems to be in tiled mode, so when you're resizing the app to be bigger (which is still al little janky) instead of being black, white or whatever was there before, the top and/or right edge is a copy of the other edge side of the window.

Now resizing the app to be bigger,  area not covered will be white.

**Review by wmww — APPROVED** [review]

**Review by wmww — APPROVED** [review]
LGTM

File: shell/platform/linux/fl_backing_store_provider.h
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.h:30]: Nit: missing full stops.

File: shell/platform/linux/fl_backing_store_provider.h
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.h:35]: Nit: missing full stop.

File: shell/platform/linux/fl_backing_store_provider.h
Comment (robert-ancell) [inline shell/platform/linux/fl_backing_store_provider.h:96]: Lower case in define - doesn't match `#ifdef`

**Review by robert-ancell — APPROVED** [review]
LGTM, thanks @huanghongxun!

**Comment by huanghongxun** [comment]
Maybe we need your approval? @stuartmorgan @cbracken

**Comment by stuartmorgan** [comment]
@robert-ancell is a committer, so his review is sufficient unless there's a reason someone else needs to review it.

**Pull request merged**
